aten/src/ATen/core/TensorImpl.h

```diff
@@ -110,6 +111,23 @@ struct CAFFE2_API TensorImpl : public c10::intrusive_ptr_target {
 
   explicit TensorImpl(at::Storage storage) : storage_(std::move(storage)), storage_offset_(0) {
     data_type_ = storage_ ? storage_.dtype() : caffe2::TypeMeta{};
+    if (storage_) {
```

Comment:
this looks fine to merge but IMO would should do things like:
1) take storage_offset for constructors taking Storage
2) Have a TensorTypeId for every device type we care about.
3) Take is_variable(?), TensorTypeId (?) for this as well.

The logic being this is an internal type, so the callers have to know what they want anyway, so put the complexity where it belongs (the weird constructors), not on the internal type.